### PR TITLE
Add alignment to cuda_malloc_async_memory_resource.

### PR DESCRIPTION
**Pull request by mzient**
## Category:
**Bug fix**
**New feature**

## Description:
`cudaMallocAsync` allocates memory aligned to a multiple of 256 bytes, however, DALI's memory_resource interface can be used to request overaligned memory. Before this PR, DALI would neither return properly aligned memory nor raise an error, should cudaMallocAsync return insufficiently aligned pointer.
This PR adds overalignment support by requesting larger blocks for allocations with alignment >256B and aligning the pointer accoringly. The mapping aligned->original pointer is kept (in a global state) and upon deletion, the pointer being deleted is looked up in the map and, if found, replaced with the original pointer.

Additionally, a tiny performance bug is fixed in AsyncPool tests.

### Affected modules and functionalities:
`cuda_malloc_async_memory_resource`
AsyncPool tests.

### Tests:

- [ ] Existing tests apply
- [X] ~New tests added~ Test adjusted
  - [ ] Python tests
  - [X] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [X] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [ ] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

File: dali/core/mm/malloc_resource.cc [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+namespace dali {
+namespace mm {
+
+#if CUDA_VERSION >= 11020
+
+namespace {
+
+class aligned_alloc_helper {
+ public:
+  static constexpr size_t kCudaMallocAlignment = 256;
```
Comment (JanuszL) [inline dali/core/mm/malloc_resource.cc:32]: Is it given once for all or may change depending on the version?

File: dali/core/mm/malloc_resource.cc [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+  void *alloc(size_t size, size_t align, cudaStream_t stream) {
+    void *ptr;
+    if (align > kCudaMallocAlignment) {
+      CUDA_CALL(cudaMallocAsync(&ptr, size + align - kCudaMallocAlignment, stream));
+      if (detail::is_aligned(ptr, size))
```
Comment (JanuszL) [inline dali/core/mm/malloc_resource.cc:38]:
```suggestion
      if (detail::is_aligned(ptr, align))
```

File: dali/core/mm/malloc_resource.cc [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+ private:
+  spinlock mtx_;
+  std::unordered_map<void *, void*> orig_ptrs_;
+};
+
+}  // namespace
+
+cuda_malloc_async_memory_resource::cuda_malloc_async_memory_resource(int device_id) {
+  // Construct the helper.
+  // The `instance` construction is finished before that of the first instance of
```
Comment (JanuszL) [inline dali/core/mm/malloc_resource.cc:78]:
```suggestion
  // The `instance` construction is finished before that of the first instance of
```
I'm not sure if I understand that sentence.

Comment (mzient) [reply dali/core/mm/malloc_resource.cc:78]: I'll rephrase.

File: dali/core/mm/malloc_resource.cc [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+  static constexpr size_t kCudaMallocAlignment = 256;
+
+  void *alloc(size_t size, size_t align, cudaStream_t stream) {
+    void *ptr;
+    if (align > kCudaMallocAlignment) {
+      CUDA_CALL(cudaMallocAsync(&ptr, size + align - kCudaMallocAlignment, stream));
+      if (detail::is_aligned(ptr, size))
+        return ptr;
+
+      void *aligned = detail::align_ptr(ptr, size);
```
[the remark this answers is not in the document]
Comment (mzient) [reply dali/core/mm/malloc_resource.cc]: Done (it was done before, pushing the changes failed).

File: dali/core/mm/malloc_resource.cc (hunk 2 above)
Comment (mzient) [reply dali/core/mm/malloc_resource.cc]: Done.

File: dali/core/mm/malloc_resource.cc (hunk 1 above)
Comment (mzient) [reply dali/core/mm/malloc_resource.cc:32]: I think it's given - I asked about it on CUDA channel and all functions that allocate global memory return data aligned to at least 256B.

**Review by JanuszL — APPROVED** [review]

**Review by banasraf — APPROVED** [review]

**Pull request merged**